Rotate map in spherical harmonics space

zonca opened the pull request:
* Added support in `hp.rotate_alm` for rotation matrices, originally healpy only wrapped the version of the function with angles (this was reported in #399)
* Renamed the `rotate_map` function to `rotate_map_pixel` as it is working in pixel space
* Added new function `rotate_map_alms` that does the rotation in harmonics space, see discussion in #460 

See an example: https://gist.github.com/zonca/67a0fec459de5b9bed007f1ec071a4b5

@ziotom78 would you like to provide some feedback?

mreineck commented on healpy/rotator.py:
```diff
@@ -373,9 +377,44 @@ def angle_ref(self, *args, **kwds):
+    def rotate_map_pixel(self, m):
+        """Rotate a HEALPix map to a new reference frame in pixel space
 
+        It is always better to rotate in spherical harmonics space, see
```
Well, strictly speaking no. If you have a non-bandlimited map, both methods will give bad results, and at least the pixel based method will be faster ;)

zonca replied:
what about:

        It is generally better to rotate in spherical harmonics space, see
        the rotate_map_alms method. A case where pixel space rotation is                                                            
        better is for heavily masked maps where the spherical harmonics
        transform is not well defined.

mreineck replied:
It's fine. I'm just worried by the fact that spherical harmonic analysis is not well defined for practically any map in the world, and nobody seems to really care.

I've heard so many people say: "I've run a map through map2alm and then through alm2map and the result is really different from the input." This is _normal_ on the sphere.

Imagine you have a map containing point sources. If you rotate it using rotate_map_alms, you get tons of ringing artifacts ... probably not what you want. And increasing resolution and lmax won't help either.

zonca replied:
The docstrings should be targeted to the less experienced users, so we should give suggestions that work most of the time. An expert user should know how to go beyond that independently.
I think also mentioning point sources would be useful.

```
    It is generally better to rotate in spherical harmonics space, see
    the rotate_map_alms method. A case where pixel space rotation might be                                                          
    more suiltable is for heavily masked maps where the spherical harmonics
    transform is not well defined or maps with point sources where the spherical harmonics transform introduces ringing artifacts.
```

mreineck commented on healpy/src/_sphtools.pyx:
```diff
@@ -535,21 +547,30 @@ def rotate_alm(alm not None, double psi, double theta, double phi, lmax=None,
     if len(alm) not in (1, 3):
         for a in alm:
-            rotate_alm(a, psi, theta, phi)
+            if matrix is None:
```
I'm confused by the different ways to call rotate_alm(). Does cython allow C++-like overloading?

zonca replied:
yes

DanielLenz commented:
Thanks, I’m currently traveling and will review in 1-2 days.

> On Sep 10, 2018, at 3:15 PM, Andrea Zonca <notifications@github.com> wrote:
> 
> @zonca requested your review on: healpy/healpy#489 Rotate map in spherical harmonics space.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.

zonca commented on healpy/rotator.py:
```diff
@@ -373,9 +377,44 @@ def angle_ref(self, *args, **kwds):
+    def rotate_alm(self, alm, lmax=None, mmax=None):
+        """Rotate Alm to a new reference frame
+
+        see hp.rotate_alm, this function **returns** the rotated alms,
```
tried it with 4eb4f49, feedback to improve welcome!

zonca commented on healpy/rotator.py:
```diff
@@ -373,9 +377,44 @@ def angle_ref(self, *args, **kwds):
+        ----------
+        m : np.ndarray
+            Input map, 1 map is considered I, 2 maps:[Q,U], 3 maps:[I,Q,U]
+        use_pixel_weights : bool, default True
```
ok a3c9cdf

zonca commented on healpy/test/test_rotator.py:
```diff
@@ -35,6 +35,27 @@ def test_rotate_map_polarization():
 
+def test_rotate_map_polarization_alms():
+    lmax = 64
+    path = os.path.dirname(os.path.realpath(__file__))
```
good idea, https://github.com/healpy/healpy/issues/504

zonca commented on healpy/rotator.py:
```diff
@@ -373,9 +377,44 @@ def angle_ref(self, *args, **kwds):
+
+    def rotate_map_alms(self, m, use_pixel_weights=True, lmax=None, mmax=None):
+        """Rotate a HEALPix map to a new reference frame in spherical harmonics space
+
```
60ef062 but I'm not completely sure

zonca commented:
rebased on current master

zonca commented:
@mreineck @hivon 

I have implemented these 2 functions to do map rotation in pixel and harmonics space, this is a highly requested feature of `healpy`.

I've added docstrings to guide users (especially beginners) on how to use them.
Unfortunately my understanding of spherical transforms is not as good as I wish it to be, so, can you please read them and suggest how to improve them?

* https://github.com/healpy/healpy/pull/489/files#diff-3f667f005c02aecef2145109ad30d2d0R380
* https://github.com/healpy/healpy/pull/489/files#diff-3f667f005c02aecef2145109ad30d2d0R420

zonca commented:
@mreineck @hivon I'm merging this but please send me feedback and I'll incorporate it in the next release.

mreineck commented:
I think the documentation is sufficient. At the same time I'm still certain that the functions will be used in inappropriate ways ... this is really expert-only stuff if you ask me, and I wouldn't try to make it easy to use for beginners :)

zonca commented:
we could even push it to the extreme and make it harder to do for beginners! ;)

mreineck commented:
I wouldn't hesitate to write something like:

Only use this for visualization purposes (and even then you should be able to achieve your goal in simpler ways)!
The only situations where this operation doesn't degrade the quality of the map are highly contrived and artificial.

hivon commented:
Hi Martin,

doing the (Planck) maps from the data streams is not an option available 
to everybody who
will ever use released data.
The Planck component separated maps are available in Galactic 
coordinates, and I don't think anybody will bother
redoing them in Ecliptic or Equatorial coordinates.
People needing the *diffuse* foregrounds, or the large scale CMB, in 
non-Galactic coordinates must have the
option of rotating existing maps, either in pixel or alm space, or a 
mixture of both.
You know very well that the band-limited requirement behind rotate_alm 
is also behind the totalconvolver/conviqt
pipeline, and a lot of Planck simulations (and therefore science) were 
based on the latter.

The problem is always the same: if we don't provide a numerically clean 
implementation of a scientifically imperfect facility,
with clearly stated caveats, people will implement and use something 
much worse.

Cheers,
Eric
>
> On 13/11/2018 23:24, mreineck wrote:
>>
>> To be honest, during the whole Planck mission I've never encountered 
>> a situation where this kind or rotation was the proper thing to do.
>> If you wanted ecliptic maps, you ran the mapmaker in ecliptic 
>> coordinates. If you wanted galactic maps, you ran it in galactic 
>> coordinates. Doing a Healpix rotation and afterwards doing science 
>> with the rotated map was almost definitely not a good idea.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub 
>> <https://github.com/healpy/healpy/pull/489#issuecomment-438461464>, 
>> or mute the thread 
>> <https://github.com/notifications/unsubscribe-auth/AFVYDf4wuy2PIkKeoCDuPg7bQjZ0kMn0ks5uu0aIgaJpZM4WiRMP>.
>>
>

-- 
***********************************************************************
Eric HIVON                              Email: hivon@iap.fr

IAP                                     Tel: +33 1 44 32 81 56
98bis Blvd Arago
F-75014 Paris                           Fax: +33 1 44 32 80 01
FRANCE
***********************************************************************
